File: my_problems/day40-bit-manipulations-2/q1_singleNumberIII.js
/****
 * Single Number III
 * 
 * Problem Description

Given an array of numbers A , in which exactly two elements appear only once and all the other elements appear exactly twice. Find the two elements that appear only once.

Note: Output array must be sorted.

Problem Constraints
2 <= |A| <= 100000
1 <= A[i] <= 10^9

Input Format
First argument is an array of integer of size N.

Output Format
Return an array of two integers that appear only once.

Example Input
Input 1:
A = [1, 2, 3, 1, 2, 4]
Input 2:
A = [1, 2]

Example Output
Output 1:
[3, 4]
Output 2:
[1, 2]

Example Explanation
Explanation 1:
 3 and 4 appear only once.
Explanation 2:
 1 and 2 appear only once.
 */

/**
 * Find position of set bit
 */
const getSetBitPos = (input) => {
    let pos = 0;
    for (let i = 0; i < 32; i++) {
        if (1 & (input >> i)) {
            pos = i;
            break;
        }
    }
    return pos;
}

/**
 * Find 2 non repeating numbers from array where all other elements are repeating twice
 * @param {Array} A array of integers
 * @returns 2 non repeating numbers
 */
const singleNumberIII = (A) => {

    const N = A.length;

    // step 1: take xor of all elements in array A
    let xorOfAll = 0;
    for (let i = 0; i < N; i++) {
        xorOfAll = xorOfAll ^ A[i];
    }

    // step 2: find any set bit position
    const pos = getSetBitPos(xorOfAll);

    // step 3: based on set bit position, form 2 auxiliry buckets
    let missingNo1 = 0;
    let missingNo2 = 0;

    for (let i = 0; i < N; i++) {
        // find the bit at position pos
        if (((A[i] >> pos) & 1) === 1) {
            // set bit
            missingNo1 = missingNo1 ^ A[i];
        } else {
            // unset bit
            missingNo2 = missingNo2 ^ A[i];
        }
    }

    // sort the output array before returning
    const output = [missingNo1, missingNo2].sort((a, b) => a - b);
    console.log(output);
    return output;
}

singleNumberIII([186, 256, 102, 377, 186, 377]);

File: my_problems/day07-bit-manipulations-2/q1_Single_Number_III.java
import java.util.Arrays;

/* Single Number III */
/* Given an array of positive integers A, two integers appear only once, and all the other integers appear twice.
Find the two integers that appear only once */
public class q1_Single_Number_III {
    public int[] solve(int[] A) {

        // calculate xor of all array elements
        int xor_all = 0;
        for (int i = 0; i < A.length; i++) {
            xor_all ^= A[i];
        }

        // find any of the set bit in xor value
        int pos = 0;
        for (int i = 0; i < 32; i++) {
            if (checkBit(xor_all, i)) {
                pos = i;
                break;
            }
        }

        // form 1 groups based on set bit at pos
        int val1 = 0;
        int val2 = 0;
        for (int i = 0; i < A.length; i++) {
            if (checkBit(A[i], pos)) {
                val1 ^= A[i];
            } else {
                val2 ^= A[i];
            }
        }
        int[] arr = new int[] { val1, val2 };
        Arrays.sort(arr);
        return arr;
    }

    private boolean checkBit(int n, int pos) {
        if (((n >> pos) & 1) == 1)
            return true;
        return false;
    }

    public static void main(String[] args) {
        q1_Single_Number_III t1 = new q1_Single_Number_III();
        int[] A, out;
        {
            A = new int[] { 1, 2, 3, 1, 2, 4 };
            out = t1.solve(A);
            ArrayUtils.printArray(out);
        }
        {
            A = new int[] { 1, 2 };
            out = t1.solve(A);
            ArrayUtils.printArray(out);
        }
        {
            A = new int[] { 186, 256, 102, 377, 186, 377 };
            out = t1.solve(A);
            ArrayUtils.printArray(out);
        }
    }
}
